### ISSUE TITLE
[Feature Request] Streamable metric support

### ISSUE THREAD

**Issue by isaacgerg** [fchollet/keras#6323]
Proposal to add streamable metrics to keras:

1. Streamable metrics allow for metrics that are non-sensical averaged over batches E.g.. prob(false alarm) or prob(true positive) are usually non-sensical to be averaged over batches and are undefined for a batch that doesn't contain a positive example.

2. Ability to reset streamable metrics at epoch level.  This would allow one to compute the bulk metric on all the y_true, y_pred's each epoch on validation set and then reset for the next epoch.

3. Clearly, streamable metrics will blow up some usecases where the output is large.  This is not a one size fits all request, just a request for multi-class or float output.  E.g. binary classifier, softmax classifier, scalar/vector prediction.
